[알린] Store step5 PR

jinios opened the pull request:
step5 PR 보냅니다! :)

jinios committed:
Chore: Modify to ignore ATS

jinios committed:
Feat: Make event badge labels #6

- Make custom UILabel class: RoundedEdgeLabel
- Delete badges variable and add stackView for implement badgeStack

jinios committed:
Feat: Set delivery, discount badge enums

jinios committed:
Feat: Add Section class #7
- Model for handling sections in a tableView

jinios committed:
Test: Add UnitTest file

jinios committed:
Feat: Modify init() #7
- Use singleton: URLSession, datatask

jinios committed:
Feat: Get data from URL and reload tableView #7

- StoreItems: Add methods to receive data from url and set property
- Category: Add url property

jinios committed:
Add UnitTest class

jinios committed:
Feat: Reload table view by the notification #7

- Add NotificationCenter observer for update section when receiving data
from model
- Modify tableView datasource methods

jinios committed:
Refactor: Simple refactoring

jinios committed:
Docs: Update README -step5

godrm commented on StoreApp/StoreApp/Model/StoreItems.swift:
```diff
@@ -8,25 +8,66 @@
+                return
+            }
+        }.resume()
+    }
```
여기서 네트워크 처리하는 부분과 StoreItems 데이터 관리하는 객체를 분리하는건 어떨까요? 
데이터 관리 객체가 다 하고 있는 것처럼 보이는데 어떻게 생각하나요?

godrm commented on StoreApp/StoreApp/Model/StoreItems.swift:
```diff
@@ -8,25 +8,66 @@
+        URLSession.shared.dataTask(with: category.url) { [self, category] (data, response, error) in
+            if let error = error {
+                print("DataTask error: \(error)\n")
+                return
```
여기서 error 인 경우 아래 response 가 200 아닌 경우는 로그만 찍고 있는데
실패한 경우에 사용자가 인식하도록 흐름 제어를 하는게 좋습니다. 그럴려면 어떻게 하는게 좋을까요?

godrm commented on StoreApp/StoreApp/Model/StoreItems.swift:
```diff
@@ -55,3 +96,9 @@ struct ItemData: Codable {
 
+extension Dictionary where Key == Category {
+    func firstKey() -> Category {
+        return self.keys.first!
```
사전이 비어있을때도 동작하는 코드일까요?

godrm commented on StoreApp/StoreApp/StoreViewController.swift:
```diff
@@ -15,40 +15,66 @@ class StoreViewController: UIViewController {
 
     func numberOfSections(in tableView: UITableView) -> Int {
-        return StoreItems.categories.count
+        guard !self.storeItems.isEmpty() else { return 0 }
```
여기 ` guard !self.storeItems.isEmpty() else { return 0 }` 반복적으로 나오는데
꼭 필요한 코드던가요? 어떤 상황에서 이 예외처리로 빠지게 되던가요?

jinios replied:
이전에 초기 테이블뷰를 세팅할때는 빈 모델을 가지고 그려야해서 isEmpty() 검사가 필요했었는데, 지금은 모델이 먼저 3개의 섹션을 가지고있어서 해당 코드가 필요없습니다. 수정하겠습니다!

jinios committed:
Refactor: Delete isEmpty() #8

jinios committed:
Refactor: Simplify & Clarify codes

jinios committed:
Refactor: Split data setting logic from StoreItems #8

- Add DataSetter class for setting StoreItems
- Add CategoryProtocol
- Rename method - makeStoreItemsFromJSON

jinios commented:
피드백 주신 부분 수정했습니다!
- StoreItems객체를 setting하는 DataSetter클래스를 만들었습니다.
- 네트워크 에러일경우 json파일로 테이블뷰의 데이터객체를 만드는 방법으로 수정했습니다.

godrm commented on StoreApp/StoreApp/View/RoundedEdgeLabel.swift:
```diff
@@ -0,0 +1,64 @@
+            return self
+        case .Event:
+            self.text = withText.rawValue
+            self.backgroundColor = UIColor(r: 255, g: 211, b: 79)
```
참고로 자주쓰는 Color를 Assets 추가해서 사용하는 방법도 있습니다. 
이렇게 코드로 넣는것보다 정밀도가 높은 P3 Color를 한꺼번에 표현할 수 있어서 좋습니다. 참고하세요.
